scipy/sparse/_sputils.py

```diff
@@ -297,7 +301,8 @@ def validateaxis(axis) -> None:
         raise ValueError("axis out of range")
 
 
-def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+def check_shape(args, current_shape=None, *, allow_1d=False,
```

Comment:
Here's another option for combining `allow_1d` with `allow_nd`: pass the keyword argument `allowed_ndims=(2,)` to represent the set of valid dimensionalities.

So by default, we'd have (2,) for only 2d shapes, but CSR and DOK could pass `(1, 2)` and COO could pass `range(1, 65)` (or some large upper limit).

This should hopefully simplify the checking logic, and also makes it explicit which values are allowed at the call-site.

Reply:
Addressed at 0ea748b. Docstrings edited at https://github.com/scipy/scipy/pull/21197/commits/e568d568647deac8aa65d751b0424bcb7822f316.